src/algebra/lie/free.lean

```diff
@@ -0,0 +1,227 @@
+`ρ : free_lie_algebra R X →ₗ⁅R⁆ free_algebra R X`, where `free_algebra R X` is the free unital,
+associative algebra, regarded as a Lie algebra via the ring commutator.
+
+By uniqueness, the image of `ρ` is the Lie subalgebra generated by `X` in the free algebra, i.e.:
+`lie_subalgebra.lie_span R (free_algebra R X) (set.range (free_algebra.ι R))`. If `ρ` is injective
+we could thus take the Lie subalgebra generated by `X` in the free associative algebra as the
+definition of the free Lie algebra, and indeed some authors do this. This is valid at least over a
+field of characteristic zero but I do not believe `ρ` is injective for general `R`. Indeed the only
+proof I know that `ρ` is injective relies on using the injectivity of the map from a Lie algebra to
+its universal enveloping algebra, which does not hold in general. Furthermore, even when a Lie
+algebra does inject into its universal enveloping algebra, the only proofs I know of this fact
+use the Poincaré–Birkhoff–Witt theorem. Some related MathOverflow questions are
+[this one](https://mathoverflow.net/questions/61954/) and
+[this one](https://mathoverflow.net/questions/114701/).
+
+## Tags
+
+lie algebra, free algebra, non-unital, non-associative, universal property, forgetful functor,
+adjoint functor
+-/
+
+universes u v w
+
+noncomputable theory
+
+variables (R : Type u) (X : Type v) [comm_ring R]
+
+/-- We save characters by using Bourbaki's name `lib` for `free_non_unital_non_assoc_algebra` in
+this file. -/
+local notation `lib` := free_non_unital_non_assoc_algebra
+local notation `lib.lift` := free_non_unital_non_assoc_algebra.lift
+local notation `lib.of` := free_non_unital_non_assoc_algebra.of
+local notation `lib.lift_of_apply` := free_non_unital_non_assoc_algebra.lift_of_apply
+local notation `lib.lift_comp_of` := free_non_unital_non_assoc_algebra.lift_comp_of
+
+namespace free_lie_algebra
+
+/-- The quotient of `lib R X` by the equivalence relation generated by this relation will give us
+the free Lie algebra. -/
+inductive rel : lib R X → lib R X → Prop
+| lie_self (a : lib R X) : rel (a * a) 0
+| leibniz_lie (a b c : lib R X) : rel (a * (b * c)) (((a * b) * c) + (b * (a * c)))
+| smul (t : R) (a b : lib R X) : rel a b → rel (t • a) (t • b)
+| add (a b c : lib R X) : rel a b → rel (a + c) (b + c)
+| mul_left (a b c : lib R X) : rel b c → rel (a * b) (a * c)
+| mul_right (a b c : lib R X) : rel a b → rel (a * c) (b * c)
+
+end free_lie_algebra
+
+/-- The free Lie algebra on the type `X` with coefficients in the commutative ring `R`. -/
+@[derive inhabited]
+def free_lie_algebra := quot (free_lie_algebra.rel R X)
+
+namespace free_lie_algebra
+
+instance : has_scalar R (free_lie_algebra R X) :=
+{ smul := λ t a, quot.lift_on a (λ x, quot.mk _ (t • x)) (λ a b h, quot.sound (rel.smul t a b h)), }
+
+instance : has_add (free_lie_algebra R X) :=
+{ add            := quot.map₂ (+)
+  (λ a b c, by { rw [add_comm _ b, add_comm _ c], exact rel.add b c a }) (λ a b c, rel.add a b c), }
```

Comment:
For some of the other constructions, we renamed `| add` to `add_right` in the inductive constructors above, and then extracted this proof to a `lemma rel.add_left`. I think it would be nice to do the same thing here.

Reply:
Will do.

Reply:
Similarly for `rel.neg` which you can then probably use in `add_left_neg`.